File: ml/src/data/jolpica.py
"""Thin client for the Jolpica F1 REST API (the Ergast successor).

Each public ``get_*`` function returns a tidy pandas DataFrame for one season.
The client handles pagination (the API caps ``limit`` at 100), reuses a single
HTTP session, and pauses briefly between calls to be polite to the free API.

Run nothing here directly — these functions are called by the caching layer
(``cache.py``) and ``scripts/fetch_data.py``.
"""
from __future__ import annotations

import time

import pandas as pd
import requests

BASE_URL = "https://api.jolpi.ca/ergast/f1"
PAGE_SIZE = 100          # API maximum
MIN_INTERVAL = 0.5       # min seconds between ANY two requests (Jolpica burst limit ~4/s)
MAX_RETRIES = 6          # retries on 429 / 5xx
TIMEOUT = 30

_session = requests.Session()
_session.headers.update({"User-Agent": "f1-guidebook/0.1 (portfolio project)"})
_last_call = [0.0]       # monotonic timestamp of the previous request


def _throttle() -> None:
    """Space requests out so we stay under the API's per-second burst limit."""
    elapsed = time.monotonic() - _last_call[0]
    if elapsed < MIN_INTERVAL:
        time.sleep(MIN_INTERVAL - elapsed)
    _last_call[0] = time.monotonic()


def _get(path: str, params: dict) -> dict:
    """GET {BASE_URL}/{path}.json with throttling + retry on 429/5xx."""
    url = f"{BASE_URL}/{path}.json"
    for attempt in range(MAX_RETRIES):
        _throttle()
        resp = _session.get(url, params=params, timeout=TIMEOUT)
        if resp.status_code == 429 or resp.status_code >= 500:
            # respect Retry-After if present, else exponential backoff
            wait = float(resp.headers.get("Retry-After", 2 ** attempt))
            time.sleep(min(wait, 30))
            continue
        resp.raise_for_status()
        return resp.json()
    resp.raise_for_status()  # retries exhausted — surface the error
    return resp.json()


def _iter_pages(path: str, params: dict | None = None):
    """Yield each page's ``MRData`` dict, looping ``offset`` until all rows fetched."""
    params = dict(params or {})
    offset = 0
    while True:
        params.update(limit=PAGE_SIZE, offset=offset)
        mrdata = _get(path, params)["MRData"]
        yield mrdata
        total = int(mrdata.get("total", 0))
        offset += PAGE_SIZE
        if offset >= total:
            break


def _coerce_numeric(df: pd.DataFrame, cols: list[str]) -> pd.DataFrame:
    for c in cols:
        if c in df.columns:
            df[c] = pd.to_numeric(df[c], errors="coerce")
    return df


def get_race_results(season: int) -> pd.DataFrame:
    """One row per driver per race: finishing position, grid, status, points."""
    rows = []
    for mr in _iter_pages(f"{season}/results"):
        for race in mr["RaceTable"]["Races"]:
            meta = {
                "season": int(race["season"]),
                "round": int(race["round"]),
                "race_name": race["raceName"],
                "circuit_id": race["Circuit"]["circuitId"],
                "date": race.get("date"),
            }
            for res in race.get("Results", []):
                drv, con = res["Driver"], res["Constructor"]
                rows.append({
                    **meta,
                    "driver_id": drv["driverId"],
                    "code": drv.get("code"),
                    "driver_name": f'{drv["givenName"]} {drv["familyName"]}',
                    "nationality": drv.get("nationality"),
                    "constructor_id": con["constructorId"],
                    "constructor_name": con["name"],
                    "grid": res.get("grid"),
                    "position": res.get("position"),       # classified finishing position
                    "position_text": res.get("positionText"),  # 'R', 'D', etc. for non-finishes
                    "points": res.get("points"),
                    "laps": res.get("laps"),
                    "status": res.get("status"),           # 'Finished', 'Accident', '+1 Lap', ...
                })
    return _coerce_numeric(pd.DataFrame(rows), ["grid", "position", "points", "laps"])


def get_qualifying(season: int) -> pd.DataFrame:
    """One row per driver per race with Q1/Q2/Q3 lap-time strings (e.g. '1:29.708')."""
    rows = []
    for mr in _iter_pages(f"{season}/qualifying"):
        for race in mr["RaceTable"]["Races"]:
            meta = {"season": int(race["season"]), "round": int(race["round"])}
            for q in race.get("QualifyingResults", []):
                drv, con = q["Driver"], q["Constructor"]
                rows.append({
                    **meta,
                    "driver_id": drv["driverId"],
                    "code": drv.get("code"),
                    "constructor_id": con["constructorId"],
                    "quali_position": q.get("position"),
                    "q1": q.get("Q1"),
                    "q2": q.get("Q2"),
                    "q3": q.get("Q3"),
                })
    return _coerce_numeric(pd.DataFrame(rows), ["quali_position"])


def get_driver_standings(season: int) -> pd.DataFrame:
    """End-of-season driver standings: position, points, wins, team."""
    rows = []
    for mr in _iter_pages(f"{season}/driverStandings"):
        for sl in mr["StandingsTable"]["StandingsLists"]:
            for s in sl.get("DriverStandings", []):
                drv = s["Driver"]
                cons = s.get("Constructors", [])
                rows.append({
                    "season": int(sl["season"]),
                    "round": int(sl["round"]),
                    "driver_id": drv["driverId"],
                    "code": drv.get("code"),
                    "position": s.get("position"),
                    "points": s.get("points"),
                    "wins": s.get("wins"),
                    "constructor_id": cons[0]["constructorId"] if cons else None,
                })
    return _coerce_numeric(pd.DataFrame(rows), ["position", "points", "wins"])


def get_constructor_standings(season: int) -> pd.DataFrame:
    """End-of-season constructor standings: position, points, wins."""
    rows = []
    for mr in _iter_pages(f"{season}/constructorStandings"):
        for sl in mr["StandingsTable"]["StandingsLists"]:
            for s in sl.get("ConstructorStandings", []):
                con = s["Constructor"]
                rows.append({
                    "season": int(sl["season"]),
                    "round": int(sl["round"]),
                    "constructor_id": con["constructorId"],
                    "constructor_name": con["name"],
                    "nationality": con.get("nationality"),
                    "position": s.get("position"),
                    "points": s.get("points"),
                    "wins": s.get("wins"),
                })
    return _coerce_numeric(pd.DataFrame(rows), ["position", "points", "wins"])


def get_schedule(season: int) -> pd.DataFrame:
    """The season calendar: round, race name, circuit, date, location."""
    rows = []
    for mr in _iter_pages(f"{season}"):
        for race in mr["RaceTable"]["Races"]:
            loc = race["Circuit"].get("Location", {})
            rows.append({
                "season": int(race["season"]),
                "round": int(race["round"]),
                "race_name": race["raceName"],
                "circuit_id": race["Circuit"]["circuitId"],
                "circuit_name": race["Circuit"]["circuitName"],
                "country": loc.get("country"),
                "locality": loc.get("locality"),
                "date": race.get("date"),
            })
    return pd.DataFrame(rows)


# --- small helper used later by feature engineering (quali_time_gap_to_pole) ---
def laptime_to_seconds(text: str | None) -> float | None:
    """Convert a lap-time string like '1:29.708' or '58.402' to float seconds."""
    if not text or not isinstance(text, str):
        return None
    text = text.strip()
    try:
        if ":" in text:
            minutes, rest = text.split(":", 1)
            return int(minutes) * 60 + float(rest)
        return float(text)
    except (ValueError, TypeError):
        return None

File: web/lib/jolpica.ts
// Live F1 data from Jolpica with ISR + graceful static fallback.
import { teams, drivers, races, getTeam } from "./data";
import type {
  ConstructorStanding,
  DriverStanding,
  ScheduleRace,
  ErgastResponse,
  SeasonDriverStat,
  DriverCareer,
} from "./types";

const BASE = "https://api.jolpi.ca/ergast/f1";
const REVALIDATE_SECONDS = 3600;
const HOUR = 3600;
const DAY = 86400;

// Politeness gate: ensure a minimum spacing between Jolpica requests so the
// sync script (and any runtime caller) never bursts past the rate limit.
const MIN_INTERVAL_MS = 500;
let gateChain: Promise<void> = Promise.resolve();
function throttle(): Promise<void> {
  const wait = gateChain.then(() => new Promise<void>((r) => setTimeout(r, MIN_INTERVAL_MS)));
  gateChain = wait;
  return gateChain;
}

async function jget(
  path: string,
  revalidate = REVALIDATE_SECONDS,
  tries = 5
): Promise<ErgastResponse> {
  for (let attempt = 0; ; attempt++) {
    await throttle();
    const res = await fetch(`${BASE}/${path}`, { next: { revalidate } });
    if (res.ok) return (await res.json()) as ErgastResponse;
    // Retry transient rate-limit (429) / server errors with backoff.
    if ((res.status === 429 || res.status >= 500) && attempt < tries - 1) {
      await new Promise((r) => setTimeout(r, 400 * (attempt + 1)));
      continue;
    }
    throw new Error(`Jolpica ${path} -> ${res.status}`);
  }
}

/**
 * Seconds until career data should refresh — just after the next race finishes,
 * since whole-career stats only change on race day. Off-season (no upcoming
 * race) falls back to ~7 days. Clamped to [1h, 30d].
 */
export async function getCareerRevalidate(season = "current"): Promise<number> {
  try {
    const { data } = await getSchedule(season);
    const now = Date.now();
    const next = data
      .map((r) => new Date(`${r.date}T${r.time ?? "14:00:00Z"}`).getTime() + 4 * HOUR * 1000)
      .filter((end) => end > now)
      .sort((a, b) => a - b)[0];
    if (!next) return 7 * DAY;
    const secs = Math.round((next - now) / 1000);
    return Math.min(Math.max(secs, HOUR), 30 * DAY);
  } catch {
    return DAY;
  }
}

export async function getConstructorStandings(
  season = "current"
): Promise<{ data: ConstructorStanding[]; live: boolean }> {
  try {
    const json = await jget(`${season}/constructorstandings/`);
    const list = json.MRData.StandingsTable?.StandingsLists?.[0]?.ConstructorStandings ?? [];
    if (!list.length) throw new Error("empty");
    return {
      live: true,
      data: list.map((c) => ({
        position: Number(c.position),
        constructorId: c.Constructor.constructorId,
        name: c.Constructor.name,
        points: Number(c.points),
        wins: Number(c.wins),
      })),
    };
  } catch {
    return {
      live: false,
      data: [...teams]
        .sort((a, b) => b.points - a.points)
        .map((t, i) => ({
          position: i + 1,
          constructorId: t.id,
          name: t.name,
          points: t.points,
          wins: t.wins,
        })),
    };
  }
}

export async function getDriverStandings(
  season = "current"
): Promise<{ data: DriverStanding[]; live: boolean }> {
  try {
    const json = await jget(`${season}/driverstandings/`);
    const list = json.MRData.StandingsTable?.StandingsLists?.[0]?.DriverStandings ?? [];
    if (!list.length) throw new Error("empty");
    return {
      live: true,
      data: list.map((d) => ({
        position: Number(d.position),
        driverId: d.Driver.driverId,
        code: d.Driver.code ?? "",
        givenName: d.Driver.givenName,
        familyName: d.Driver.familyName,
        nationality: d.Driver.nationality,
        points: Number(d.points),
        wins: Number(d.wins),
        constructorId: d.Constructors[0]?.constructorId ?? "",
        constructorName: d.Constructors[0]?.name ?? "",
      })),
    };
  } catch {
    return {
      live: false,
      data: [...drivers]
        .sort((a, b) => b.points - a.points)
        .map((d, i) => {
          const [givenName, ...rest] = d.name.split(" ");
          return {
            position: i + 1,
            driverId: d.id,
            code: d.code,
            givenName,
            familyName: rest.join(" "),
            nationality: d.nationality,
            points: d.points,
            wins: d.wins,
            constructorId: d.team,
            constructorName: getTeam(d.team)?.name ?? d.team,
          };
        }),
    };
  }
}

/**
 * Aggregates real per-driver season stats (races, wins, podiums, fastest laps
 * from race results; poles from qualifying) for the given season, keyed by the
 * driver's 3-letter code. A few cached, season-wide calls shared by every team page.
 */
export async function getSeasonStats(
  season = "current"
): Promise<{ data: Record<string, SeasonDriverStat>; live: boolean }> {
  try {
    const stats: Record<string, SeasonDriverStat> = {};
    const slot = (code: string) =>
      (stats[code] ??= { races: 0, wins: 0, podiums: 0, poles: 0, fastestLaps: 0 });

    // Race results across every round (paginated) -> races / wins / podiums / fastest laps.
    const limit = 100;
    let offset = 0;
    let total = Infinity;
    while (offset < total) {
      const json = await jget(`${season}/results/?limit=${limit}&offset=${offset}`);
      total = Number(json.MRData.total ?? 0);
      const racesPage = json.MRData.RaceTable?.Races ?? [];
      if (!racesPage.length) break;
      for (const r of racesPage) {
        for (const res of r.Results ?? []) {
          const code = res.Driver.code;
          if (!code) continue;
          const s = slot(code);
          s.races += 1;
          const pos = Number(res.position);
          if (pos === 1) s.wins += 1;
          if (pos <= 3) s.podiums += 1;
          if (res.FastestLap?.rank === "1") s.fastestLaps += 1;
        }
      }
      offset += limit;
    }

    // Poles = qualifying P1 per round.
    const q = await jget(`${season}/qualifying/1/?limit=100`);
    for (const r of q.MRData.RaceTable?.Races ?? []) {
      const code = r.QualifyingResults?.[0]?.Driver.code;
      if (code) slot(code).poles += 1;
    }

    if (!Object.keys(stats).length) throw new Error("empty");
    return { live: true, data: stats };
  } catch {
    return { live: false, data: {} };
  }
}

/**
 * Whole-career totals for a driver (by Ergast driverId). Each value is the
 * `total` count of a filtered results/qualifying query, fetched in parallel.
 * Podiums = finishes P1 + P2 + P3.
 */
export async function getDriverCareer(
  driverId: string,
  revalidate?: number
): Promise<{ data: DriverCareer | null; live: boolean }> {
  try {
    const totalOf = async (path: string) =>
      Number((await jget(path, revalidate)).MRData.total ?? 0);
    // Serial (not Promise.all) to avoid bursting Jolpica's rate limit.
    const races = await totalOf(`drivers/${driverId}/results/?limit=1`);
    const p1 = await totalOf(`drivers/${driverId}/results/1/?limit=1`);
    const p2 = await totalOf(`drivers/${driverId}/results/2/?limit=1`);
    const p3 = await totalOf(`drivers/${driverId}/results/3/?limit=1`);
    const poles = await totalOf(`drivers/${driverId}/qualifying/1/?limit=1`);
    const fastestLaps = await totalOf(`drivers/${driverId}/fastest/1/results/?limit=1`);
    return {
      live: true,
      data: { races, wins: p1, podiums: p1 + p2 + p3, poles, fastestLaps },
    };
  } catch {
    return { live: false, data: null };
  }
}

export async function getSchedule(
  season = "current"
): Promise<{ data: ScheduleRace[]; live: boolean }> {
  try {
    const json = await jget(`${season}/races/`);
    const list = json.MRData.RaceTable?.Races ?? [];
    if (!list.length) throw new Error("empty");
    return {
      live: true,
      data: list.map((r) => ({
        season: Number(r.season),
        round: Number(r.round),
        name: r.raceName,
        circuit: r.Circuit.circuitName,
        locality: r.Circuit.Location?.locality,
        country: r.Circuit.Location?.country,
        date: r.date,
        time: r.time,
      })),
    };
  } catch {
    const latestSeason = Math.max(...races.map((r) => r.season));
    return {
      live: false,
      data: races
        .filter((r) => r.season === latestSeason)
        .sort((a, b) => a.round - b.round)
        .map((r) => ({
          season: r.season,
          round: r.round,
          name: r.name,
          circuit: r.circuit,
          date: r.date,
        })),
    };
  }
}
